docs/mockchain.1
.\" mockchain
.TH "mockchain" "1" "@VERSION@" "Seth Vidal" ""
.SH "NAME"
mockchain - chain package builder.
.SH "SYNOPSIS"
\fBmockchain\fP -r chroot_config [options]  [package...]
.SH "DESCRIPTION"
.PP
\fBmockchain\fP builds a series of srpms in mock one at a time. After each 
successful build of a package it adds the resulting packages to a local 
repo which are available to the next package to satisfy buildreqs.
.SH "USAGE NOTES"
.PP
This does not try to sort the packages by build order b/c that is too much 
effort and not obviously doable with the buildreq information we have.
.PP
The build process when you use -l is idempotent so a package which has 
already been successfully built will not be built again.
.PP
If you want to force the rebuild of a package which has been built 
successfully simply remove the 'success' file from the directory for 
the package in the localrepo path. 

.SH OPTIONS
.TP
\fB\-a\fR REPOS, \fB\-\-addrepo\fR=\fIREPOS\fR
add this repo baseurl to the yumconfig for the chroot. This can
be specified multiple times. Let's you point to multiple
paths beyond the default to pull build deps from.

.TP
\fB\-c\fR, \fB\-\-continue\fR
if a pkg fails to build, continue to the next one, default is to stop

.TP
\fB\-l\fR LOCALREPO, \fB\-\-localrepo\fR=\fIPATH\fR
set the path to put the results/repo in. This path needs to be
somewhere accessible to users other than you for reading as the
mock process doesn't run as you. Will make a tempdir if not set.

.TP
\fB\-\-log\fR=\fILOGFILE\fR
log to the file named by this option, defaults to not
logging

.TP
\fB\-m\fR OPTION, \fB\-\-mock-option\fR=\fIOPTION\fR
pass the OPTION to mock. Can be used several times. For example:

\fB\-m \-\-define="packager Monkey"\fR

.TP
\fB\-r\fR CONFIG, \fB\-\-root\fR=\fICONFIG\fR
chroot config name to use in the mock build

.TP
\fB\-\-recurse\fR
build all pkgs, record the failures and try to rebuild them
again and again until everything gets built (or until the 
set of pkgs failing to build are the same over) sets --continue

.SH "AUTHORS"
Seth Vidal <skvidal@fedoraproject.org>


ChangeLog
2007-12-01  Michael Brown <mebrown@michaels-house.net>
	* configure.ac
	remove unnecessary configure items to speed configure
	* daemontest.c
	fix compile warnings
	* doc/releasetests.sh
	more comprehensive release tests
	* decorator.py
	remove our decorator.py in favor of std python-decoratortools
	* various
	pylint/pychecker recommendations
	make decorator syntax compatible with python 2.3 (for RHEL4 porting...)
	fix 'mock install'
	* mock.py
	popagate chroot/shell command exit codes to caller
	init groups list if we are called by sudo so we have access to
		all directories caller does
	

2007-12-01  Michael Brown <mebrown@michaels-house.net>
	* backend.py
	fixed build of multiple srpms on one cmdline

2007-11-29  Clark Williams  <williams@redhat.com>

	* mock.py, Makefile* 
	added back verbosity control options (-q and -v)
	added --offline option
	added dependency on python-cytypes
	lots of logging changes

	
2007-10-20  Michael Brown <mebrown@michaels-house.net>

	* mock.py
	lots of work redoing stuff. Now use setuid wrapper instead of
	setuid helper.
	huge number of changes upstream
	convert to setuid wrapper instead of old setuid helper
	lots of bugfixes and improvements
	/var/cache/yum now saved and bind-mounted
	ccache integration
	rootcache improvements (formerly called autocache)

2007-02-24  Clark Williams  <williams@redhat.com>

	* mock.spec
	BZ#226673 - Ville Skyttä fix for RPM_OPT_FLAGS

2007-02-20  Clark Williams  <williams@redhat.com>

	* mock.py
	Handle --no-clean option when doing yum.conf symlink
	
	* mock.spec 
	Bumped version to 0.6.13

2007-02-16  Clark Williams  <williams@redhat.com>

	* mock.py
	Added safety symlink to handle possible yum breakage in rawhide

2007-02-07  Clark Williams  <williams@redhat.com>

	* mock.py
	Added error() calls to print command output on failed commands

2007-02-06  Clark Williams  <williams@redhat.com>

	* mock.py
	added installdeps command for long-term chroot management
	fixed invalid reference to os.file (rather than os.path)
	
2007-01-08  Clark Williams  <williams@redhat.com>
	* Makefile
	changed archive creation to keep tarball clean

	* etc/fedora-4-*
	added Josh Boyer's EPEL config files

	* mock.spec
	bumped version to 0.6.10
	
2006-10-30  Clark Williams  <williams@redhat.com>

	* Makefile
	Bumped version to 0.6.7 for FC6 release

	* etc/fedora-6-*
	Added for FC6 release

2006-10-21  Clark Williams  <williams@redhat.com>

	* mock.spec
	Bumped version to 0.6.6

2006-09-11 11:21  jcwillia
	* mock.py
	added patch from Karanbir Singh to fix RPM problem in 
	CentOS 4.4

	* mock.spec
	bumped version to 0.6.5

2006-08-29 05:16  jcwillia

	* mock.py
	explicit import of rpmUtils.transaction 
	(BZ#204051)

2006-08-25 16:40  jcwillia

	* buildsys-build.spec
	Dennis Gilmore's patches for Aurora SPARC Linux 
	(BZ#204100)
	
2006-08-17 20:28  jcwillia
	
	* etc/redhat-73-i386.cfg, etc/redhat-9-i386.cfg:
	added runuser config option for legacy config files
	
	* mock.py
	updated usage message
	
	* doc/mock.1
	reworked man page to include caching options, chroot and shell
	commands. 
	(BZ#200080)

2006-08-17 20:23  jcwillia
	
	* etc/redhat-8-i386.cfg:
	added template for RHL8
	
2006-08-16 16:59  jcwillia

	* mock.spec
	added disttags
	
	* mock.py
	added check for basedir change and raise error when it's detected
	(BZ#1644441)
	
	
2006-08-07 17:26  jcwillia
	
	* buildsys-build.spec
	added rhel tags

	* Makefile
	Modified to build all supported targets for buildsys-build

2006-08-07 15:05  jcwillia

	* buildsys-build.spec
	Thorsten Leemhuis and Paul Howarth's patches for RHL and Legacy
	chroots.
	(BZ#196930)
	
2006-06-13 01:31  skvidal

	* etc/defaults.cfg: 
	add cache defaults

2006-06-13 01:16  skvidal

	* ChangeLog: 
	check in changelog

2006-06-13 01:15  skvidal

	* etc/defaults.cfg: 
	correct commented-out groupinstall option

2006-06-13 00:28  skvidal

	* etc/: fedora-3-i386-core.cfg, fedora-3-x86_64-core.cfg,
	fedora-4-i386-core.cfg, fedora-4-ppc-core.cfg,
	fedora-4-x86_64-core.cfg, fedora-5-x86_64-core.cfg,
	fedora-devel-i386-core.cfg, fedora-devel-ppc-core.cfg,
	fedora-devel-x86_64-core.cfg: 
	config file changes to point to mirrors

2006-06-13 00:12  skvidal

	* etc/: fedora-4-i386-core.cfg, fedora-4-ppc-core.cfg,
	fedora-4-x86_64-core.cfg, fedora-5-i386-core.cfg,
	fedora-5-ppc-core.cfg, fedora-5-x86_64-core.cfg: 
	fix up some of the url/mirrorlist paths

2006-06-12 23:52  skvidal

	* buildsys-build.spec: 
	reduce build set

2006-06-08 17:28  mebrown

	* mock.py, src/mock-helper.c: Make rm less verbose.

2006-06-07 08:36  skvidal

	* etc/: fedora-1-i386-core.cfg, fedora-2-i386-core.cfg,
	fedora-3-i386-core.cfg, fedora-3-x86_64-core.cfg,
	fedora-4-i386-core.cfg, fedora-4-ppc-core.cfg,
	fedora-4-x86_64-core.cfg, fedora-5-i386-core.cfg,
	fedora-5-ppc-core.cfg, fedora-5-x86_64-core.cfg,
	fedora-devel-i386-core.cfg, fedora-devel-ppc-core.cfg,
	fedora-devel-x86_64-core.cfg, redhat-73-i386.cfg,
	redhat-9-i386.cfg: 
	fix up all the paths to buildgroups and to plague build-results

2006-06-07 08:29  skvidal

	* mock.py, mock.spec: 
	
	iterate version in spec files

2006-06-07 08:27  skvidal

	* etc/: fedora-1-i386-core.cfg, fedora-2-i386-core.cfg,
	fedora-3-i386-core.cfg, fedora-3-x86_64-core.cfg,
	fedora-4-i386-core.cfg, fedora-4-ppc-core.cfg,
	fedora-4-x86_64-core.cfg, fedora-5-i386-core.cfg,
	fedora-5-ppc-core.cfg, fedora-5-x86_64-core.cfg,
	fedora-devel-i386-core.cfg, fedora-devel-ppc-core.cfg,
	fedora-devel-x86_64-core.cfg, redhat-73-i386.cfg,
	redhat-9-i386.cfg: 
	clean out things that are now in 'defaults.cfg'

2006-06-06 17:28  skvidal

	* etc/defaults.cfg: use defaults more intelligently


2006-05-24 11:15  skvidal

	* mock.py, src/mock-helper.c: 
	check in Michael Brown's lovely autocache creation patch.

2006-05-17 11:12  jcwillia

	* mock.py: Michael E. Brown's patch for getting PS1 down to chroot
	shell

2006-05-16 00:15  skvidal

	* mock.py: applied tab-fix and a slimming of main patches from
	Michael E. Brown.


py/mockchain.py
#!/usr/bin/python -tt
# by skvidal@fedoraproject.org
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Library General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301  USA.
# copyright 2012 Red Hat, Inc.

# SUMMARY
# mockchain
# take a mock config and a series of srpms
# rebuild them one at a time
# adding each to a local repo
# so they are available as build deps to next pkg being built
from __future__ import print_function
try:
    from six.moves.urllib_parse import urlsplit
except ImportError:
    from urlparse import urlsplit
import sys
import subprocess
import os
import optparse
import requests
import tempfile
import shutil
import time
import re

import mockbuild.util

# all of the variables below are substituted by the build system
__VERSION__ = "unreleased_version"
SYSCONFDIR = os.path.join(os.path.dirname(os.path.realpath(sys.argv[0])), "..", "etc")
PYTHONDIR = os.path.dirname(os.path.realpath(sys.argv[0]))
PKGPYTHONDIR = os.path.join(os.path.dirname(os.path.realpath(sys.argv[0])), "mockbuild")
MOCKCONFDIR = os.path.join(SYSCONFDIR, "mock")
# end build system subs

mockconfig_path = '/etc/mock'

def createrepo(path):
    if os.path.exists(path + '/repodata/repomd.xml'):
        comm = ['/usr/bin/createrepo_c', '--update', path]
    else:
        comm = ['/usr/bin/createrepo_c', path]
    cmd = subprocess.Popen(comm,
             stdout=subprocess.PIPE, stderr=subprocess.PIPE)
    out, err = cmd.communicate()
    return out, err

def parse_args(args):
    parser = optparse.OptionParser('\nmockchain -r mockcfg pkg1 [pkg2] [pkg3]')
    parser.add_option('-r', '--root', default=None, dest='chroot',
            metavar="CONFIG",
            help="chroot config name/base to use in the mock build")
    parser.add_option('-l', '--localrepo', default=None,
            help="local path for the local repo, defaults to making its own")
    parser.add_option('-c', '--continue', default=False, action='store_true',
            dest='cont',
            help="if a pkg fails to build, continue to the next one")
    parser.add_option('-a', '--addrepo', default=[], action='append',
            dest='repos',
            help="add these repo baseurls to the chroot's yum config")
    parser.add_option('--recurse', default=False, action='store_true',
            help="if more than one pkg and it fails to build, try to build the rest and come back to it")
    parser.add_option('--log', default=None, dest='logfile',
            help="log to the file named by this option, defaults to not logging")
    parser.add_option('--tmp_prefix', default=None, dest='tmp_prefix',
            help="tmp dir prefix - will default to username-pid if not specified")
    parser.add_option('-m', '--mock-option', default=[], action='append',
            dest='mock_option',
            help="option to pass directly to mock")


    opts, args = parser.parse_args(args)
    if opts.recurse:
        opts.cont = True

    if not opts.chroot:
        print("You must provide an argument to -r for the mock chroot")
        sys.exit(1)


    if len(sys.argv) < 3:
        print("You must specifiy at least 1 package to build")
        sys.exit(1)


    return opts, args

REPOS_ID = []
def generate_repo_id(baseurl):
    """ generate repository id for yum.conf out of baseurl """
    repoid = "/".join(baseurl.split('//')[1:]).replace('/', '_')
    repoid = re.sub(r'[^a-zA-Z0-9_]', '', repoid)
    suffix = ''
    i = 1
    while repoid + suffix in REPOS_ID:
        suffix = str(i)
        i += 1
    repoid = repoid + suffix
    REPOS_ID.append(repoid)
    return repoid

def add_local_repo(infile, destfile, baseurl, repoid=None):
    """take a mock chroot config and add a repo to it's yum.conf
       infile = mock chroot config file
       destfile = where to save out the result
       baseurl = baseurl of repo you wish to add"""
    global config_opts

    try:
        with open(infile) as f:
            code = compile(f.read(), infile, 'exec')
        exec(code)
        if not repoid:
            repoid = generate_repo_id(baseurl)
        else:
            REPOS_ID.append(repoid)
        localyumrepo = """
[%s]
name=%s
baseurl=%s
enabled=1
skip_if_unavailable=1
metadata_expire=30
cost=1
""" % (repoid, baseurl, baseurl)

        config_opts['yum.conf'] += localyumrepo
        br_dest = open(destfile, 'w')
        for k, v in list(config_opts.items()):
            br_dest.write("config_opts[%r] = %r\n" % (k, v))
        br_dest.close()
        return True, ''
    except (IOError, OSError):
        return False, "Could not write mock config to %s" % destfile

    return True, ''

def do_build(opts, cfg, pkg):

    # returns 0, cmd, out, err = failure
    # returns 1, cmd, out, err  = success
    # returns 2, None, None, None = already built

    s_pkg = os.path.basename(pkg)
    pdn = s_pkg.replace('.src.rpm', '')
    resdir = '%s/%s' % (opts.local_repo_dir, pdn)
    resdir = os.path.normpath(resdir)
    if not os.path.exists(resdir):
        os.makedirs(resdir)

    success_file = resdir + '/success'
    fail_file = resdir + '/fail'

    if os.path.exists(success_file):
        return 2, None, None, None

    # clean it up if we're starting over :)
    if os.path.exists(fail_file):
        os.unlink(fail_file)

    mockcmd = ['/usr/bin/mock',
               '--configdir', opts.config_path,
               '--resultdir', resdir,
               '--uniqueext', opts.uniqueext,
               '-r', cfg, ]
    # heuristic here, if user pass for mock "-d foo", but we must be care to leave
    # "-d'foo bar'" or "--define='foo bar'" as is
    compiled_re_1 = re.compile(r'^(-\S)\s+(.+)')
    compiled_re_2 = re.compile(r'^(--[^ =])[ =](\.+)')
    for option in opts.mock_option:
        r_match = compiled_re_1.match(option)
        if r_match:
            mockcmd.extend([r_match.group(1), r_match.group(2)])
        else:
            r_match = compiled_re_2.match(option)
            if r_match:
                mockcmd.extend([r_match.group(1), r_match.group(2)])
            else:
                mockcmd.append(option)

    print('building %s' % s_pkg)
    mockcmd.append(pkg)
    cmd = subprocess.Popen(mockcmd,
           stdout=subprocess.PIPE,
           stderr=subprocess.PIPE)
    out, err = cmd.communicate()
    if cmd.returncode == 0:
        open(success_file, 'w').write('done\n')
        ret = 1
    else:
        if (isinstance(err, bytes)):
            err = err.decode("utf-8")
        sys.stderr.write(err)
        open(fail_file, 'w').write('undone\n')
        ret = 0

    return ret, cmd, out, err


def log(lf, msg):
    if lf:
        now = time.time()
        try:
            open(lf, 'a').write(str(now) + ':' + msg + '\n')
        except (IOError, OSError) as e:
            print('Could not write to logfile %s - %s' % (lf, str(e)))
    print(msg)


config_opts = {}

def main(args):
    opts, args = parse_args(args)
    # take mock config + list of pkgs
    cfg = opts.chroot
    pkgs = args[1:]

    global config_opts
    config_opts = mockbuild.util.load_config(mockconfig_path, cfg, None, __VERSION__, PKGPYTHONDIR)

    if not opts.tmp_prefix:
        try:
            opts.tmp_prefix = os.getlogin()
        except OSError as e:
            print("Could not find login name for tmp dir prefix add --tmp_prefix")
            sys.exit(1)
    pid = os.getpid()
    opts.uniqueext = '%s-%s' % (opts.tmp_prefix, pid)


    # create a tempdir for our local info
    if opts.localrepo:
        local_tmp_dir = os.path.abspath(opts.localrepo)
        if not os.path.exists(local_tmp_dir):
            os.makedirs(local_tmp_dir)
            os.chmod(local_tmp_dir, 0o755)
    else:
        pre = 'mock-chain-%s-' % opts.uniqueext
        local_tmp_dir = tempfile.mkdtemp(prefix=pre, dir='/var/tmp')
        os.chmod(local_tmp_dir, 0o755)

    if opts.logfile:
        opts.logfile = os.path.join(local_tmp_dir, opts.logfile)
        if os.path.exists(opts.logfile):
            os.unlink(opts.logfile)

    log(opts.logfile, "starting logfile: %s" % opts.logfile)
    opts.local_repo_dir = os.path.normpath(local_tmp_dir + '/results/' + config_opts['chroot_name'] + '/')

    if not os.path.exists(opts.local_repo_dir):
        os.makedirs(opts.local_repo_dir, mode=0o755)

    local_baseurl = "file://%s" % opts.local_repo_dir
    log(opts.logfile, "results dir: %s" % opts.local_repo_dir)
    opts.config_path = os.path.normpath(local_tmp_dir + '/configs/' + config_opts['chroot_name'] + '/')

    if not os.path.exists(opts.config_path):
        os.makedirs(opts.config_path, mode=0o755)

    log(opts.logfile, "config dir: %s" % opts.config_path)

    my_mock_config = os.path.join(opts.config_path, "{0}.cfg".format(config_opts['chroot_name']))

    # modify with localrepo
    res, msg = add_local_repo(config_opts['config_file'], my_mock_config, local_baseurl, 'local_build_repo')
    if not res:
        log(opts.logfile, "Error: Could not write out local config: %s" % msg)
        sys.exit(1)

    for baseurl in opts.repos:
        res, msg = add_local_repo(my_mock_config, my_mock_config, baseurl)
        if not res:
            log(opts.logfile, "Error: Could not add: %s to yum config in mock chroot: %s" % (baseurl, msg))
            sys.exit(1)


    # these files needed from the mock.config dir to make mock run
    for fn in ['site-defaults.cfg', 'logging.ini']:
        pth = mockconfig_path + '/' + fn
        shutil.copyfile(pth, opts.config_path + '/' + fn)


    # createrepo on it
    out, err = createrepo(opts.local_repo_dir)
    if err.strip():
        log(opts.logfile, "Error making local repo: %s" % opts.local_repo_dir)
        log(opts.logfile, "Err: %s" % err)
        sys.exit(1)


    download_dir = tempfile.mkdtemp()
    downloaded_pkgs = {}
    built_pkgs = []
    try_again = True
    to_be_built = pkgs
    return_code = 0
    num_of_tries = 0
    while try_again:
        num_of_tries += 1
        failed = []
        for pkg in to_be_built:
            if not pkg.endswith('.rpm'):
                log(opts.logfile, "%s doesn't appear to be an rpm - skipping" % pkg)
                failed.append(pkg)
                continue

            elif pkg.startswith('http://') or pkg.startswith('https://') or pkg.startswith('ftp://'):
                url = pkg
                try:
                    log(opts.logfile, 'Fetching %s' % url)
                    r = requests.get(url)
                    if r.status_code == requests.codes.ok:
                        fn = urlsplit(r.url).path.rsplit('/', 1)[1]
                        pkg = download_dir + '/' + fn
                        fd = open(pkg, 'wb')
                        for chunk in r.iter_content(4096):
                            fd.write(chunk)
                        fd.close()
                except Exception as e:
                    log(opts.logfile, 'Error Downloading %s: %s' % (url, str(e)))
                    failed.append(url)
                    continue
                else:
                    downloaded_pkgs[pkg] = url
            log(opts.logfile, "Start build: %s" % pkg)
            ret, cmd, out, err = do_build(opts, config_opts['chroot_name'], pkg)
            log(opts.logfile, "End build: %s" % pkg)
            if ret == 0:
                failed.append(pkg)
                log(opts.logfile, "Error building %s." % os.path.basename(pkg))
                if opts.recurse:
                    log(opts.logfile, "Will try to build again (if some other package will succeed).")
                else:
                    log(opts.logfile, "See logs/results in %s" % opts.local_repo_dir)
            elif ret == 1:
                log(opts.logfile, "Success building %s" % os.path.basename(pkg))
                built_pkgs.append(pkg)
                # createrepo with the new pkgs
                out, err = createrepo(opts.local_repo_dir)
                if err.strip():
                    log(opts.logfile, "Error making local repo: %s" % opts.local_repo_dir)
                    log(opts.logfile, "Err: %s" % err)
            elif ret == 2:
                log(opts.logfile, "Skipping already built pkg %s" % os.path.basename(pkg))

        if failed and opts.recurse:
            if len(failed) != len(to_be_built):
                to_be_built = failed
                try_again = True
                log(opts.logfile, 'Some package succeeded, some failed.')
                log(opts.logfile, 'Trying to rebuild %s failed pkgs, because --recurse is set.' % len(failed))
            else:
                log(opts.logfile, "Tried %s times - following pkgs could not be successfully built:" % num_of_tries)
                for pkg in failed:
                    msg = pkg
                    if pkg in downloaded_pkgs:
                        msg = downloaded_pkgs[pkg]
                    log(opts.logfile, msg)
                try_again = False
        else:
            try_again = False
            if failed:
                return_code = 2

    # cleaning up our download dir
    shutil.rmtree(download_dir, ignore_errors=True)

    log(opts.logfile, "Results out to: %s" % opts.local_repo_dir)
    log(opts.logfile, "Pkgs built: %s" % len(built_pkgs))
    if built_pkgs:
        if failed:
            if len(built_pkgs):
                log(opts.logfile, "Some packages successfully built in this order:")
        else:
            log(opts.logfile, "Packages successfully built in this order:")
        for pkg in built_pkgs:
            log(opts.logfile, pkg)
    return return_code

if __name__ == "__main__":
    sys.exit(main(sys.argv))
